File: Sliding Window/Frequency of the Most Frequent Element/solution_binary_search.ts
function maxFrequency(nums: number[], k: number): number {
    nums.sort((a,b) => a-b)
    let prefix = Array(nums.length).fill(0)
    prefix[0] = nums[0]
    for(let i = 1; i < nums.length; i++) {
        prefix[i] = prefix[i-1] + nums[i]
    }
    let result = 1
    for(let i = 1; i < nums.length; i++) {
        let target = nums[i] * (i + 1)
        let sum = prefix[i]
        let operations = target - sum
        if(operations <= k) {
            result = Math.max(result, i + 1)
            continue
        }
        let min = 0
        let max = i
        while(min < max) {
            let mid = min + Math.floor((max - min) / 2)
            let sum = prefix[i] - prefix[mid]
            let target = nums[i] * (i - (mid+1) + 1)
            let operations = target - sum
            if(operations > k) {
                min = mid + 1
            } else {
                max = mid
            }
        }
        result = Math.max(result, i - (min+1) + 1)
    }
    return result
};

File: Array/*Minimum Cost to Make Array Equal/solution_binary_search.py
class Solution:
    def minCost(self, nums: List[int], cost: List[int]) -> int:
        n = len(nums)
        start, end = min(nums), max(nums)
        result = sum([abs(nums[i] - start) * cost[i] for i in range(n)])
        while start < end:
            mid = start + (end - start) // 2
            c = sum([abs(mid - nums[i]) * cost[i] for i in range(n)])
            c_next = sum([abs(mid + 1 - nums[i]) * cost[i] for i in range(n)])
            if c < c_next:
                end = mid
            # in case it's equal, we can think of it as 2 medians situations
            # it will have only up to 2 medians. so the answer's moving to the right.
            else:
                start = mid + 1
            result = min(result, c, c_next)
        return result
